clippy_lints/src/utils/hir_utils.rs

```diff
@@ -195,7 +195,10 @@ impl<'a, 'tcx> SpanlessEq<'a, 'tcx> {
                 ls == rs && over(l, r, |l, r| self.eq_pat(l, r))
             },
             (&PatKind::Range(ref ls, ref le, ref li), &PatKind::Range(ref rs, ref re, ref ri)) => {
-                self.eq_expr(ls, rs) && self.eq_expr(le, re) && (*li == *ri)
+                if let (Some(ls), Some(rs), Some(le), Some(re)) = (ls, rs, le, re) {
```

Comment:
Not sure why you're not supporting half-open ranges here.

Reply:
We should also support some patterns like `(Some(ls), None, None, Some(re))` here?

Reply:
Well, surely e.g., `1..` and `1..` should compare equal, so you need something like:
```rust
both(ls, rs, |a, b| self.eq_expr(a, b))
&& both(le, re, |a, b| self.eq_expr(a, b))
&& (*li == *ri)
```